### ISSUE TITLE
Fail to send from Google AppEngine but works locally

### ISSUE THREAD

**Issue by allankellynet** [sendgrid/docs#2499]
I'm slowly going mad here.. I'm working in Python 2.7 on the Google AppEngine, I've got some code (below) which works perfectly well when running locally - in the Google AppEngine Launcher. It correctly sends mail through SendGrid and I can see the mail in the activity log.

But, when I deploy the code to AppEngine for real it fails with an HTTP 400 Error which suggest some authorization failure. The SendGrid docs that are references discuss browser based usage which doesn't make sense
- is it something to authorization of the Google domain?
- or sender e-mail addresses? (that does differ between the local config and live)

Code is:
   def post_one(self, conference, msg, submission_key):
        self.sg = sendgrid.SendGridAPIClient(apikey=os.environ.get('SENDGRID_API_KEY'))

        submission = submission_key.get()
        mail = sendgrid.helpers.mail.Mail(
                    sendgrid.helpers.mail.Email(conference.contact_email(), conference.name),
                    msg.build_subject(submission),
                    sendgrid.helpers.mail.Email(submission.talk.get().key.parent().get().email),
                    sendgrid.helpers.mail.Content("text/plain", msg.build_message(submission)))

        for cc in massage_none_list(msg.cc_addresses(submission_key)):
            mail.personalizations[0].add_cc(sendgrid.helpers.mail.Email(cc))

        for bcc in massage_none_list(msg.bcc_addresses(submission_key)):
            mail.personalizations[0].add_bcc(sendgrid.helpers.mail.Email(bcc))

        if sysinfo.is_running_local():
            logging.info("Mail sening to: " +
                         submission.talk.get().key.parent().get().email)

        try:
            response = self.sg.client.mail.send.post(request_body=mail.get())

            # SendGrid error messages https://sendgrid.com/docs/API_Reference/Web_API_v3/Mail/errors.html
            if response.status_code >= 200 and response.status_code <= 300:
                msg.mark_comms(submission)
            else:
                logging.error("Mail failure sending to: " + " Response status code: " + response.status_code)
                logging.info(mail.get())
                msg.mark_comms_fail(submission)
        except urllib2.HTTPError, e:
            logging.error("Mail failure with HTTPError exception.")
            logging.info(mail.get())
            logging.info(e.code)
            logging.info(e.reason)
            logging.info(e.headers)
            msg.mark_comms_fail(submission)

And the AppEngine log (with some text sanitized):
16:23:12.072 Mail failure with HTTPError exception.
16:23:12.073 {'content': [{'type': 'text/plain', 'value': u'Re...'}], 'personalizations': [{'cc': [{'email': u'...@gmail.com'}, {'email': u'...@....co.uk'}, {'email': u'...@....co.uk'}, {'email': u'...@gmail.com'}], 'to': [{'email': u'...@gmail.com'}]}], 'from': {'name': u'... 2017', 'email': u's@...com'}, 'subject': u'... 2017'}
16:23:12.144 400
16:23:12.145 Bad Request
16:23:12.145 Server: nginx
Date: Thu, 23 Feb 2017 16:23:12 GMT
Content-Type: application/json
Content-Length: 349
Connection: keep-alive
X-Frame-Options: DENY
Access-Control-Allow-Origin: https://sendgrid.api-docs.io
Access-Control-Allow-Methods: POST
Access-Control-Allow-Headers: Authorization, Content-Type, On-behalf-of, x-sg-elas-acl
Access-Control-Max-Age: 600
X-No-CORS-Reason: https://sendgrid.com/docs/Classroom/Basics/API/cors.html

**Comment by thinkingserious** [comment]
Hello @allankellynet,

In this function: `os.environ.get('SENDGRID_API_KEY')`, we are expecting the variable SENDGRID_API_KEY to be set in the environment. I'm guessing this is not happening properly on App Engine.

This looks like it might be what you need: https://cloud.google.com/appengine/docs/standard/python/config/appref

With Best Regards,

Elmer

**Comment by allankellynet** [comment]
Sorry, no. 
From the app.yaml
env_variables:
    SENDGRID_API_KEY: "..."

At one point I had some logging code in there to make sure the key was picked up, and it was.
And before that I had the key hard coded.

**Comment by thinkingserious** [comment]
Hi @allankellynet,

Can you please log the following so we can discover the source of the "BAD REQUEST"?

https://github.com/sendgrid/sendgrid-python/blob/master/TROUBLESHOOTING.md#error-messages

We need to figure out what the "HTTPError exception" is.

**Comment by allankellynet** [comment]
Yippee! - that nailed it.

For the recorded, I got...

{"errors": [{"message":
   "Each email address in the personalization block should be unique between to, cc, and bcc. We found the first duplicate instance of [...@gmail.com] in the personalizations.0.cc field.","field":"personalizations.0","help":"http://sendgrid.com/docs/API_Reference/Web_API_v3/Mail/errors.html#message.recipient-errors"}]}

I was testing with my own Gmail account but I was included on the CC list as one of the admins.

Many thanks
